### ISSUE TITLE
Is there any mechanism for followers to send feedback: like, dislike, comment, share?

### ISSUE THREAD

**Issue by ratpoison4** [fiatjaf/nostr#1]
I guess share can be implemented using just republishing the post to your followers.

But how to implement likes, dislikes and comments?

**Comment by fiatjaf** [comment]
Everything is just events published to relays. You can put anything in events, therefore you can publish "like" events and the reader can see these. I think it's quite wasteful to publish "like" events, but it can be done and probably will be done, so we'll document and standardize event codes for these things.

But we also can come up up with less wasteful ways of handling these things, I have some ideas.

**Comment by ratpoison4** [comment]
> You can put anything in events, therefore you can publish "like" events

Who are you referring as "you" in this sentence? The original publisher or the person who makes like/dislike/comment?

We need to think about some protection against manipulation with likes. If it is the publisher who is expected to publish "like" events, he can increase likes and decrease dislikes. Some kind of protection is needed to make sure that likes come from real people.

Also in case of comments it is even more important to authenticate the comment, so that the author of the comment signed it. Also the publisher needs to sign as well probably for it to appear under the post.

Such actions are 3 way interactions, because they involve not only a publisher and a reader, but also a 3rd party (liker or commentator) which should probably be identified and authenticated as well as the publisher.

We may think also about anonymity of likes/comments but with the proof that the author is real. May be done with blind signatures probably. Liker presents a like or dislike to the publisher, he signs without knowing is it like or dislike (or the contents of the comment), then the liker removes blindness and publishes it to the servers. If the stream of likes/dislikes is large, then the publisher can not figure out who made what but he knows that a person did something (like or dislike).

**Comment by fiatjaf** [comment]
The only things that exist are events. All events are signed. An event can be anything. It can be a post, it can be a like to a post.

All events are published to relays, so there is no chance to do this blind signatures stuff because there is no direct interaction between publisher and liker (of course you could come up with something using relays but that would cause complexity to go to the roof so it is not desirable).

There is no guarantee you'll ever see all events. There is no guarantee a relay will have all events you may want. And that's the solution, I think.

To prevent antispam a relay can implement their own techniques for filtering out likes from keys that they consider spam (which can be many things). And the clients  can filter out too according to their own criteria (but the easiest way is to filter out according to the relay from which these comments come).

**Comment by ratpoison4** [comment]
IIUC clients download content produced by public keys they follow. But likes and comments come not from guys they follow, but from other readers. How does the client know that the post he is reading has likes/comments associated with it? He need to watch the post probably and download not only posts from people he follows, but also events associated with posts he read.

**Comment by elazar** [comment]
Have you looked at or considering adapting any of these protocols instead of writing one from scratch?

* [Activity Streams 2.0](https://www.w3.org/TR/activitystreams-core/) / [Activity Vocabulary](https://www.w3.org/TR/activitystreams-vocabulary/)
* [ActivityPub](https://www.w3.org/TR/activitypub/) (without the federated server-to-server component)
* [WebSub](https://www.w3.org/TR/websub/)
* [Micropub](https://www.w3.org/TR/micropub/#source-content)

**Comment by fiatjaf** [comment]
@elazar I did consider ActivityPub and the reasons it wouldn't work are explained in this repository's README (in short: the crucial problem is that you do not own your identity).

I've never heard about these others, I will take a look at them later today, but can you tell me if they are being used by someone or some app somewhere?

**Comment by elazar** [comment]
> I've never heard about these others, I will take a look at them later today, but can you tell me if they are being used by someone or some app somewhere?

According to https://activitystrea.ms:

> The Activity Streams format has already been adopted by BBC, IBM, Superfeedr, TypePad, and [many others](http://wiki.activitystrea.ms/w/page/24500522/Implementors).

WebSub has [implementation reports](https://github.com/w3c/websub/tree/master/implementation-reports) that appear to include Mastodon, Twitch, and others.

Micropub has [implementation reports](https://github.com/w3c/Micropub/tree/master/implementation-reports) as well, though they don't include any household names that I can see.

**Comment by fiatjaf** [comment]
> The Micropub protocol is used to create, update and delete posts on one's own domain using third-party clients.

So no, we don't want that. Although I very much like servers and people using their own domains and websites, this is not the goal here, we don't want a protocol people have to run servers to be able to use. We want something that can really be used by everyone and not a small group of enthusiasts.

**Comment by elazar** [comment]
> So no, we don't want that. Although I very much like servers and people using their own domains and websites, this is not the goal here, we don't want a protocol people have to run servers to be able to use. We want something that can really be used by everyone and not a small group of enthusiasts.

Do you mean "a protocol *everyone has* to run servers to be able to use?" Because relays as described in NIP-01 are functioning as servers, are they not? So at least some users have to run a server for this protocol to work? Otherwise, I'd assume you'd be angling for a peer-to-peer solution.

In any case, the stated purpose of Micropub from its standard doesn't have to align directly with the same for this project in order for this project to use the data format described in the standard. I included Micropub mainly because it's a lightweight format and seems to have a lot in common with what's described in NIP-01.

**Comment by fiatjaf** [comment]
> WebSub provides a common mechanism for communication between publishers of any kind of Web content and their subscribers, based on HTTP web hooks. Subscription requests are relayed through hubs, which validate and verify the request. Hubs then distribute new and updated content to subscribers when it becomes available. WebSub was previously known as PubSubHubbub.

So this is a very generic and flexible protocol for writing webhook hubs that can relay anything, it seems.

Also doesn't fit the Nostr basic idea, because Nostr clients cannot receive webhooks. It serves more to setups in which each people is running their own server and want to communicate between these servers.

Could be used by an auxiliary protocol that allowed different Nostr relays to share data among them (but if you're implementing something specific already that isn't going to benefit at all from the other existing WebSub implementations then I think it doesn't matter if you reuse their protocol or come up with a new one).

**Comment by fiatjaf** [comment]
> Do you mean "a protocol everyone has to run servers to be able to use?" Because relays as described in NIP-01 are functioning as servers, are they not? So at least some users have to run a server for this protocol to work? Otherwise, I'd assume you'd be angling for a peer-to-peer solution.

Yes, I mean that. Yes, we need relays in Nostr, which are servers, but they don't have any identity of their own, their URL is not meaningful, they're not seem by users except when users go to manually edit their relay configurations, they're just invisible and dumb hubs.

> In any case, the stated purpose of Micropub from its standard doesn't have to align directly with the same for this project in order for this project to use the data format described in the standard. I included Micropub mainly because it's a lightweight format and seems to have a lot in common with what's described in NIP-01.

Indeed you have a point, it has some similarities with the Nostr ideas here. I think maybe we could use this protocol, let me look more into it.

**Comment by fiatjaf** [comment]
I thought it might be a good idea to reuse Micropub so we could be integrated from the start with multiple clients capable of publishing notes, but this hypothesis has proved false, since Micropub is made with a different purpose and the clients expect you to be tied to one domain already from the start -- which is contrary to the Nostr idea of publishing to multiple relays. Also most clients are served from a server and they're not standalone clients. Plus you cannot in any way sign your notes like Nostr clients are expected to do, and that's an essential Nostr feature.

I did found one single example of a standalone Micropub that could be repurposed (hopefully with not much effort) to be a Nostr client: https://omnibear.com/

I will look into doing that now and see if it works.

**Comment by cristiano-belloni** [comment]
For likes, relays could accept likes and store them in RAM, then debounce and aggregate them. If a post gets a lot of likes in a short time span, a relay can generate a single message with "likes: 1200" (for example) linked to that post after 10 minutes, instead of 1200 single like messages.

**Comment by fiatjaf** [comment]
That's a good idea, but who would sign that message? The relay directly could have a keypair of its own and people would trust the like-count from that pubkey or not.

**Comment by cristiano-belloni** [comment]
> That's a good idea, but who would sign that message? The relay directly could have a keypair of its own and people would trust the like-count from that pubkey or not.

Yeah, that adds to the complexity and makes relays not completely interchangeable. But the problem of decentralized social network is that regular users want feature-parity with established solutions.

**Comment by librehodl** [comment]
Let's just bite the bullet and publish the like event.

And if that gets done. Adding the follow event would be hyper helpful.

Letting users know who is following them will make them not feel like shouting into the void.

**Comment by scsibug** [comment]
Do we not already have follows from NIP-02?

I am skeptical of likes as a standalone event.  A client may have to download hundreds of KB of data for a single well-liked event (say, 1,000 likes).  And it is just asking for spammers to create fake accts w/ likes.  Only requesting like events from your followers may alleviate the spam problem, but having to follow people to see your likes is not ideal either.

I suspect that like events, if implemented, will be a great use case for relays to implement event retention configurable by event kind.  I myself would probably drop `like` events 30 days after they were created.  Enough for clients to see recent likes, but with some limits to growth.  Curious what users/clients would think about this behavior (think 'recently liked' instead of 'ever liked').  That also, hopefully, discourages people from publishing delete events against their old likes!

**Comment by librehodl** [comment]
The ephemeral thing is very smart. 

The p2p reddit clone from @ethereans also only records like 2 weeks of event data (on server) and 6 months of data (on client) https://getaether.net/docs/tech/

I think they prevent spam by using HashCash proof of work.

**Comment by fiatjaf** [comment]
The best way to do likes I could come up with is by clients trusting relays to aggregate them and then the relay just sends a number (probably attached to the event being liked, for example) to clients instead of hundreds of different events.

But this or other approaches would have to be at least half-implemented first so we can say they work before we codify them into NIPs.

**Comment by wsq003** [comment]
@fiatjaf  https://github.com/bazarinitiative/BazarServer could be the `relay` you imagined. Although it implemented for another similar protocol.

**Comment by earonesty** [comment]
i think it's not possible to detect relays that are dishonest with this method.  better to just relay likes and focus on efficiency
